File: src/service/lease.service.js
const _ = require('lodash');
const NoDevicesAvailableException = require('../model/exception/no-devices-available.exception');
const LeaseNotFoundException = require('../model/exception/lease-not-found.exception');

class LeaseService {

    constructor(leaseDao, deviceDao) {
        this.leaseDao = leaseDao;
        this.deviceDao = deviceDao;
    }

    allLeases() {
        return this.leaseDao.allLeases();
    }

    leaseAny() {
        const leases = this.leaseDao.allLeases();

        const device = this.deviceDao.allDevices()
            .find(device => _.every(leases, lease => lease.device.id !== device.id));

        if (device) {
            return this.leaseDao.create(device);
        }
        throw new NoDevicesAvailableException();
    }

    release(leaseId) {
        const removed = this.leaseDao.delete(leaseId);
        if (_.isEmpty(removed)) {
            throw new LeaseNotFoundException(leaseId);
        }
    }
}

module.exports = LeaseService;

File: src/main/kotlin/org/andcoe/adf/leases/LeaseService.kt
package org.andcoe.adf.leases

import org.andcoe.adf.devices.Device
import org.andcoe.adf.devices.DeviceId
import org.andcoe.adf.devices.DeviceService
import org.andcoe.adf.exceptions.DeviceNotFound

class LeaseService(private val deviceService: DeviceService,
                   private val leaseDao: LeaseDao) {

    fun create(): Lease? {
        val leases = leaseDao.leases()
        val devices = deviceService.devices()

        val device: Device? = devices.values
            .find { device -> leases.all { it.value.device.deviceId != device.deviceId } }

        return if (device != null) leaseDao.create(device) else null
    }

    fun create(deviceId: DeviceId): Lease? {
        val devices = deviceService.devices()
        val device: Device = devices[deviceId] ?: throw DeviceNotFound("No device found with id: '${deviceId.id}'")

        val leases = leaseDao.leases()
        val leaseNotFound = leases.all { lease -> lease.value.device.deviceId != device.deviceId }
        return if (leaseNotFound) leaseDao.create(device) else null
    }
}